tests/platform_tests/fwutil/fwutil_common.py

```diff
@@ -0,0 +1,228 @@
+import pytest
+import os
+import json
+import logging
+
+from copy import deepcopy
+
+from tests.common.utilities import wait_until
+
+logger = logging.getLogger(__name__)
+
+DEVICES_PATH="usr/share/sonic/device"
+TIMEOUT=3600
+REBOOT_TYPES = {
+        "cold": "reboot",
+        "warm": "warm-reboot",
+        "fast": "fast-reboot"
+        }
+
+def find_pattern(lines, pattern):
+    for line in lines:
+        if pattern.match(line):
+            return True
+    return False
+
+def power_cycle(duthost=None, pdu_ctrl=None, delay_time=60):
+    if pdu_ctrl is None:
+        pytest.skip("No PSU controller for %s, skipping" % duthost.hostname)
+
+    all_outlets = pdu_ctrl.get_outlet_status()
+
+    for outlet in all_outlets:
+        pdu_ctrl.turn_off_outlet(outlet)
+    time.sleep(delay_time)
+    for outlet in all_outlets:
+        pdu_ctrl.turn_on_outlet(outlet)
+
+def reboot(duthost, pdu_ctrl, reboot_type):
+    if reboot_type == "power off": 
+        power_cycle(duthost, pdu_ctrl)
+        return
+
+    if reboot_type not in REBOOT_TYPES: pytest.fail("Invalid reboot type {}".format(reboot_type))
+
+    duthost.command(REBOOT_TYPES[reboot_type], module_ignore_errors=True, module_async=True)
+
+def complete_install(duthost, localhost, boot_type, res, pdu_ctrl, auto_reboot=False, current=None, next_image=None):
+    hn = duthost.mgmt_ip
+
+    if boot_type != "none":
+        if not auto_reboot:
+            logger.info("Waiting on install to finish.")
+            res.get(TIMEOUT)
+            logger.info("Rebooting switch using {} boot".format(boot_type))
+            duthost.command("sonic-installer set-default {}".format(current))
+            reboot(duthost, pdu_ctrl, boot_type)
+        
+        logger.info("Waiting on switch to shutdown...")
+        # Wait for ssh flap
+        localhost.wait_for(host=hn, port=22, state='stopped', delay=10, timeout=TIMEOUT)
```

Comment:
@alexrallen looks like there is a code duplicate for switch shutdown/startup check

Reply:
Unsure what you are referring to. If it is the code block for `# Reboot back into original image if neccesary.` that is for a firmware install which forced a reboot on us during the "next image" test and we need to reboot to get back to the image under test.

Reply:
@alexrallen you have
```python
logger.info("Waiting on switch to shutdown....")
logger.info("Waiting on switch to come up....")
```
in both places.

Also, why do you need to use:
```python
    # Reboot back into original image if neccesary
    if next_image:
        logger.info("We booted into the new image... booting back into the image under test.")
        duthost.command("sonic-installer set-default {}".format(current))
```
Why not to use `set-next-boot`?

Reply:
The log lines are duplicated but really only because they are doing the same exact thing. This is used in the case in which (like BIOS) the fwutil forces us into the next sonic image through a forced reboot before the code is able to set the next boot back to the image under test. 

On installation of the new sonic image it sets the default boot to the new one, this must be reset back to the current image.